New core functions: toDate and toDateTime

jamisonbryant opened the pull request:
Adds two new global functions: `toDate()` and `toDateTime()` which accept mixed values and attempt to convert them to the native Chronos object. This is the first part of a series of PRs to make type-safe request data available to users.

jamisonbryant committed:
Add functions for coercing to dates and timestamps

jamisonbryant committed:
Break up if / else / else if blocks

jamisonbryant committed:
Merge branch '5.next' into feature/to-date-to-datetime

jamisonbryant committed:
Remove Frozen date/time classes

(Stan can't find them because they're aliased)

markstory commented on src/Core/functions.php:
```diff
@@ -526,3 +531,88 @@ function toBool(mixed $value): ?bool
+ * Converts a value to a native DateTime object.
+ *
+ *  integer - value is treated as a Unix timestamp
+ *  string - value is treated as a ISO-8601 (Atom) formatted timestamp
+ *  Other values returns as null.
```
I like this direction. I wonder if we could also use the default formatting types that `Database\Type\DateTimeType` uses. A potentially simpler solution would be to allow users to provide a format to parse with and default the parameter to the `ATOM` format, or ISO-8601 format.

markstory commented on src/Core/functions.php:
```diff
@@ -526,3 +531,88 @@ function toBool(mixed $value): ?bool
+ * @return Date|null Returns a Date object if parsing is successful, or NULL otherwise.
+ * @since 5.1.0
+ */
+function toDate(mixed $value): ?Date
```
Same here with having a format string with a default, so that userland code can do their thing as needed, and that we could use it in `Database\Type\DateType` as well.

jamisonbryant replied:
I added a `$format` argument to both functions, can you take a quick look and verify that's what you were expecting, and that I got the format strings from the correct place?

ADmad commented on src/Core/functions.php:
```diff
@@ -526,3 +531,88 @@ function toBool(mixed $value): ?bool
+        return DateTime::parse($value);
+    }
+
+    if (is_int($value)) {
```
With the recent changes I made to [Chronos::createFromTimestamp()](https://github.com/cakephp/chronos/pull/454) it accepts floats too, so this check could be updated to allow floats. Also considering values from query strings, they would be numeric strings, so do we want to just use a `is_numeric()` check here?

jamisonbryant replied:
I've made the change to `is_numeric` but I need to update the tests now. I bumped the `cakephp/chronos` dependency to 3.1 to get @ADmad's change, just checking that was OK to do?

jamisonbryant committed:
Merge branch '5.next' into feature/to-date-to-datetime

jamisonbryant committed:
Add format arguments

jamisonbryant committed:
Change is_int to is_numeric

jamisonbryant committed:
Clarify where format is applied

jamisonbryant committed:
Bump cakephp/chronos version

jamisonbryant committed:
Cast strings to floats

markstory commented on src/Core/functions.php:
```diff
@@ -16,7 +16,14 @@
+use Exception;
+use IntlDateFormatter;
 use JsonException;
+use Opis\JsonSchema\Formats\DateTimeFormats;
```
This isn't a package we have in our dependency list 🤔

jamisonbryant replied:
Wow, not sure where that came from, I was playing around with DateTime format constants and I guess my IDE pulled it in. Anyway, it's not in use so I will remove it. Good catch, thanks.

jamisonbryant committed:
Remove unused imports

jamisonbryant committed:
Fix tests for toDateTime()

jamisonbryant committed:
Fix some tests for toDate()

jamisonbryant committed:
Fix the remainder of tests for toDate()

jamisonbryant committed:
Add link to SimpleDateFormat docs

jamisonbryant committed:
Remove one unused import

jamisonbryant committed:
Merge branch '5.next' into feature/to-date-to-datetime

jamisonbryant committed:
Switch to Date::parse()

jamisonbryant committed:
Progress on tests

ADmad commented on tests/TestCase/Core/FunctionsTest.php:
```diff
@@ -683,4 +689,142 @@ public static function toBoolProvider(): array
         ];
     }
+
+    /**
+     * @dataProvider toDateTimeProvider
+     */
```
```suggestion
    #[DataProvider('toDateTimeProvider')]
```

ADmad commented on tests/TestCase/Core/FunctionsTest.php:
```diff
@@ -683,4 +689,142 @@ public static function toBoolProvider(): array
+        ];
+    }
+
+    /**
+     * @dataProvider toDateProvider
+     */
```
```suggestion
    #[DataProvider('toDateProvider')]
```

ADmad commented on src/Core/functions.php:
```diff
@@ -16,6 +16,10 @@
  */
 namespace Cake\Core;
 
+use Cake\I18n\Date;
```
This makes the i18n package a dependency of the core package, don't really like that. Perhaps these `toDate(Time)` functions should go into `I18n/functions.php` instead.

jamisonbryant replied:
I agree that having the core depend on i18n instead of the other way around is a bit concerning, so I took a quick look at `i18n/functions.php`. All of those functions are prefixed by a double-underscore, but I think it would be unusual to repeat that pattern for these two functions (not that you suggested that in particular).

So, would it be appropriate to move these functions to the other `functions.php` but NOT do the double-underscore prefix naming convention? Or is that a hard rule?

ADmad replied:
There's no need for new functions added to the I18n/functions.php file also have underscore prefix unless they are performing a similar task :)

markstory replied:
The `__` on those functions is a convention inherited from [gettext](https://www.php.net/manual/en/function.-.php) which provides `_`. Our non-extension implementation of po/mo files uses `__` so the other text translation functions also use `__` 🤷 . However, new functions need not bother with that.

:+1: on ADmad's suggestion of putting these functions into `i18n/functions.php`. They'll still be autoloadable but keeps our internal dependencies simpler.

jamisonbryant committed:
Apply suggestions from code review

Switch to PHPUnit annotations for data providers

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

jamisonbryant committed:
Change Exception to Throwable

jamisonbryant commented:
I just realized also that some tests are failing on PHP8.4 specifically that are passing on 8.3 and below:

```
DateRangeError: DateTimeImmutable::createFromTimestamp(): Argument #1 ($timestamp) must be a finite 
  number between -9223372036854775808 and 9223372036854775807.999999, NAN given
// (same for -INF)
// (same for INF)
```

I tried work around this by simply adding `|\DateRangeError` to the try-catch block but this class is only available in PHP8.3 and up, and `composer.json` is currently set at >=8.1. So I changed the exception to catch `Throwable` instead. I've never written try-catch blocks this 'broad' before so if this the wrong approach I would love to get some feedback.

Alternatively, we can remove the tests for NAN/INF and assume users will not try to pass these values to the functions, which I think is semi-reasonable.

jamisonbryant committed:
Move functions to I18n

jamisonbryant committed:
Move tests to new test class

jamisonbryant committed:
Remove from core functions test

ADmad commented on src/I18n/functions.php:
```diff
@@ -226,3 +229,91 @@ function __dxn(
 }
+
+/**
+ * Converts a value to a native DateTime object.
```
```suggestion
 * Converts a value to a DateTime object.
```

The term "native" is generally used for PHP's classes, which isn't the case here and can be misleading.

ADmad commented on src/I18n/functions.php:
```diff
@@ -226,3 +229,91 @@ function __dxn(
+}
+
+/**
+ * Converts a value to a native Date object.
```
```suggestion
 * Converts a value to a Date object.
```

ADmad commented on src/I18n/functions.php:
```diff
@@ -226,3 +229,91 @@ function __dxn(
+ * @param string $format The date format the value is in. Defaults to Short (ex: 1970-01-01) format.
+ * @return Date|null Returns a Date object if parsing is successful, or NULL otherwise.
+ * @since 5.1.0
+ *
+ * @see https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classSimpleDateFormat.html#details
```
```suggestion
```
We don't use ICU's formatter anymore.

jamisonbryant committed:
Apply suggestions from code review

Language correction and remove link that no longer applies

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

markstory committed:
Make tests more durable to slow clocks

CI often has slow machines that can have a long delay between data
providers and tests run.

markstory committed:
Fix split package requirements
